File: NominalAttributes/nominal_dissimilarity.py
def nominal_dissimilarity(x, y):
    """
    Calculates dissimilarity for nominal attributes.
    d(i, j) = (p - m) / p
    """
    if len(x) != len(y):
        raise ValueError("Vectors must be same length")
    
    p = len(x)
    m = 0 # matches
    
    for i in range(p):
        if x[i] == y[i]:
            m += 1
            
    return (p - m) / p

# Example Data
x = ["Circle", "Red", "Small"]
y = ["Circle", "Blue", "Small"]

dist = nominal_dissimilarity(x, y)
print(f"Object x: {x}")
print(f"Object y: {y}")
print(f"Nominal Dissimilarity: {dist:.4f}")
print(f"Nominal Similarity: {1 - dist:.4f}")

File: NominalAttributes/nominal_dissimilarity.c
#include <stdio.h>
#include <string.h>

// Simple function to check string equality
int is_match(char* s1, char* s2) {
    return strcmp(s1, s2) == 0;
}

int main() {
    // Example Data
    char* x[] = {"Circle", "Red", "Small"};
    char* y[] = {"Circle", "Blue", "Small"};
    
    int p = 3; // Total attributes
    int m = 0; // Matches

    printf("Object x: {%s, %s, %s}\n", x[0], x[1], x[2]);
    printf("Object y: {%s, %s, %s}\n", y[0], y[1], y[2]);

    for (int i = 0; i < p; i++) {
        if (is_match(x[i], y[i])) {
            m++;
        }
    }

    float dissimilarity = (float)(p - m) / p;
    
    printf("Matches (m): %d\n", m);
    printf("Dissimilarity: %.4f\n", dissimilarity);
    
    return 0;
}